pkg/customunit/manager.go

```diff
@@ -0,0 +1,331 @@
+	if unitRef == "" {
+		return nil, nil
+	}
+
+	instance, err := h.getServiceInstance(ctx, ar)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get service instance from access request to fetch the api service
+	serviceRef := instance.GetReferenceByGVK(management.APIServiceGVK())
+	service := h.cache.GetAPIServiceWithName(serviceRef.Name)
+	if service == nil {
+		return nil, fmt.Errorf("could not find service connected to quota")
+	}
+	extAPIID, err := util.GetAgentDetailsValue(instance, definitions.AttrExternalAPIID)
+	if err != nil {
+		return nil, err
+	}
+
+	q := &customunits.QuotaInfo{
+		ApiInfo: &customunits.APIInfo{
+			ServiceDetails: util.GetAgentDetailStrings(service),
+			ServiceName:    service.Name,
+			ServiceID:      service.Metadata.ID,
+			ExternalAPIID:  extAPIID,
+		},
+		AppInfo: &customunits.AppInfo{
+			AppDetails: util.GetAgentDetailStrings(app),
+			AppName:    app.Name,
+			AppID:      app.Metadata.ID,
+		},
+		Quota: &customunits.Quota{
+			Count: int64(count),
+			Unit:  unitRef,
+		},
+	}
+
+	return q, nil
+}
+
+type reference struct {
+	Kind string `json:"kind"`
+	Name string `json:"name"`
+	Unit string `json:"unit"`
+}
+
+func (h *CustomUnitMetricServerManager) getQuotaInfo(ar *management.AccessRequest) (string, int) {
+	index := 0
+	if len(ar.Spec.AdditionalQuotas) < index+1 {
+		return "", 0
+	}
+
+	q := ar.Spec.AdditionalQuotas[index]
+	for _, r := range ar.References {
+		d, _ := json.Marshal(r)
+		ref := &reference{}
+		json.Unmarshal(d, ref)
+		if ref.Kind == catalog.QuotaGVK().Kind && ref.Name == q.Name {
+			return ref.Unit, int(q.Limit)
+		}
+	}
+	return "", 0
+}
+
+func (h *CustomUnitMetricServerManager) getServiceInstance(_ context.Context, ar *management.AccessRequest) (*v1.ResourceInstance, error) {
+	instRef := ar.GetReferenceByGVK(management.APIServiceInstanceGVK())
+	instID := instRef.ID
+	instance, err := h.cache.GetAPIServiceInstanceByID(instID)
+	if err != nil {
+		return nil, err
+	}
+	return instance, nil
+}
+
+func (m *CustomUnitMetricServerManager) HandleMetricReporting(ctx context.Context, cancelCtx context.CancelFunc, metricCollector metricCollector) {
+	if m.agentType != config.TraceabilityAgent {
+		return
+	}
+	go m.receiveMetrics(metricCollector)
+	// iterate over each metric service config
+	for _, config := range m.configs {
+		// Initialize custom units client
+		factory := NewCustomUnitClientFactory(config.URL, m.cache, &customunits.QuotaInfo{})
+		client, _ := factory()
+		go client.StartMetricReporting(m.metricReportChan)
+		m.clients = append(m.clients, client)
+	}
+}
+
+func (c *CustomUnitMetricServerManager) receiveMetrics(metricCollector metricCollector) {
+	for {
+		select {
+		case metricReport := <-c.metricReportChan:
+			if metricReport == nil {
+				continue
+			}
+			logger := c.logger.WithField("api", metricReport.ApiService.GetValue())
+			customMetricDetail, err := c.buildCustomMetricDetail(metricReport)
+			if err != nil {
+				logger.Error(err)
```

Comment:
you can change this to logger.WithError(err).Error("could not build metric data")